modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-evaluator-impl/src/main/java/com/liferay/dynamic/data/mapping/form/evaluator/internal/function/EqualsFunction.java

```diff
@@ -35,7 +36,43 @@
 
 	@Override
 	public Boolean apply(Object object1, Object object2) {
-		return Objects.equals(_getValue(object1), _getValue(object2));
+		Object value1 = _getValue(object1);
+
+		Object value2 = _getValue(object2);
+
+		Object object1Class = value1.getClass();
+
+		Object object2Class = value2.getClass();
+
+		if (!object1Class.equals(object2Class) &&
+			((object1 instanceof BigDecimal) ||
+			 (object2 instanceof BigDecimal))) {
+
+			if (object1 instanceof BigDecimal) {
+				if (value2 instanceof Long) {
+					value1 = GetterUtil.getLong(value1);
+				}
+				else if (value2 instanceof Double) {
+					value1 = GetterUtil.getDouble(value1);
+				}
+				else if (value2 instanceof Integer) {
+					value1 = GetterUtil.getInteger(value1);
+				}
+			}
+			else {
+				if (value1 instanceof Long) {
+					value2 = GetterUtil.getLong(value2);
+				}
+				else if (value1 instanceof Double) {
+					value2 = GetterUtil.getDouble(value2);
+				}
+				else if (value1 instanceof Integer) {
+					value2 = GetterUtil.getInteger(value2);
+				}
+			}
+		}
+
+		return Objects.equals(value1, value2);
```

Comment:
Hi @sontruongces, @thienquach87,
We can simplify this code by checking if both values ​​are instances of `Number`. If they are, we can convert them to `BigDecimal` as follows:

```
@Override
public Boolean apply(Object object1, Object object2) {
	if ((object1 instanceof Number) && (object2 instanceof Number)) {
		BigDecimal bigDecimal1 = new BigDecimal(object1.toString());
		BigDecimal bigDecimal2 = new BigDecimal(object2.toString());

		return bigDecimal1.compareTo(bigDecimal2) == 0;
	}

	return Objects.equals(_getValue(object1), _getValue(object2));
}
```
I tested it and it works, could you please double check?

Thanks!

Reply:
@carolmariaabb,
Thank you for your suggestion. However, if your code is applied, I have some concerns as below.
  1. If both object1 and object2 are the same types (one of Long, Integer or Double), we do not need to spend effort for convert to BigDecimal, and the "==" operator is used instead of equal or conpareTo function. The "==" is faster them.
  2. If both of them are BigDecimal, we do a redundance job to convert BigDecimal to BigDecimal.
  You can refer to the code below:
https://github.com/liferay/liferay-portal/blob/master/modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-evaluator-impl/src/main/java/com/liferay/dynamic/data/mapping/form/evaluator/internal/function/EqualsFunction.java#L47-L51
 and equals function 
   /**
     * Returns {@code true} if the arguments are equal to each other
     * and {@code false} otherwise.
     * Consequently, if both arguments are {@code null}, {@code true}
     * is returned and if exactly one argument is {@code null}, {@code
     * false} is returned.  Otherwise, equality is determined by using
     * the {@link Object#equals equals} method of the first
     * argument.
     *
     * @param a an object
     * @param b an object to be compared with {@code a} for equality
     * @return {@code true} if the arguments are equal to each other
     * and {@code false} otherwise
     * @see Object#equals(Object)
     */
    public static boolean equals(Object a, Object b) {
        return (a == b) || (a != null && a.equals(b));
    }

Thanks,
Thien

Reply:
Hi @thienquach87, 

You could either put the checks mentioned by you 

> If both object1 and object2 are the same types 
> If both of them are BigDecimal 

in the code suggested by me, or you could simplify your code to avoid code duplication, as follows: 

```
public Boolean apply(Object object1, Object object2) {
	Object value1 = _getValue(object1);
	Object value2 = _getValue(object2);

	if (!Objects.equals(value1.getClass(), value2.getClass())) {
		if (object1 instanceof BigDecimal) {
			value1 = _convertValue(value2, value1);
		}
		else if (object2 instanceof BigDecimal) {
			value2 = _convertValue(value1, value2);
		}
	}

	return Objects.equals(value1, value2);
}

private Object _convertValue(Object value1, Object value2) {
	if (value1 instanceof Double) {
		return GetterUtil.getDouble(value2);
	}
	else if (value1 instanceof Integer) {
		return GetterUtil.getInteger(value2);
	} 
	else if (value1 instanceof Long) {
		return GetterUtil.getLong(value2);
	}

	return value2;
}
```

Notice that I put the 

> if (value1 instanceof Double) 
> if (value1 instanceof Integer)
> if (value1 instanceof Long) 

in in alphabetic order.

Thanks!